ompi/mca/op/aarch64/op_aarch64_functions.c

```diff
@@ -0,0 +1,641 @@
+                             struct ompi_datatype_t **dtype,                      \
+                             struct ompi_op_base_module_1_0_0_t *module)          \
+    {                                                                             \
+        int types_per_step = svcnt(*((type##type_size##_t *) _in));               \
+        int left_over = *count;                                                   \
+        type##type_size##_t *in = (type##type_size##_t *) _in,                    \
+                            *out = (type##type_size##_t *) _out;                  \
+        OP_CONCAT(OMPI_OP_TYPE_PREPEND, type##type_size##_t) vsrc, vdst;          \
+        svbool_t pred = svptrue_b##type_size();                                   \
+        for (; left_over >= types_per_step; left_over -= types_per_step) {        \
+            vsrc = svld1(pred, in);                                               \
+            vdst = svld1(pred, out);                                              \
+            in += types_per_step;                                                 \
+            vdst = OP_CONCAT(OMPI_OP_OP_PREPEND, op##_x)(pred, vdst, vsrc);       \
+            OP_CONCAT(OMPI_OP_OP_PREPEND, st1)(pred, out, vdst);                  \
+            out += types_per_step;                                                \
+        }                                                                         \
+                                                                                  \
+        if (left_over > 0) {                                                      \
+            /* set the pred mask to only the remaining values */                  \
+            pred = svwhilelt_b##type_size(0, left_over);                          \
+            vsrc = svld1(pred, in);                                               \
+            vdst = svld1(pred, out);                                              \
+            vdst = OP_CONCAT(OMPI_OP_OP_PREPEND, op##_z)(pred, vdst, vsrc);       \
+            OP_CONCAT(OMPI_OP_OP_PREPEND, st1)(pred, out, vdst);                  \
+        }                                                                         \
+    }
+#endif
+
+/*************************************************************************
+ * Max
+ *************************************************************************/
+#undef current_func
+#define current_func(a, b) ((a) > (b) ? (a) : (b))
+    OP_AARCH64_FUNC(max, s,  8, 16,   int, max)
+    OP_AARCH64_FUNC(max, u,  8, 16,  uint, max)
+    OP_AARCH64_FUNC(max, s, 16,  8,   int, max)
+    OP_AARCH64_FUNC(max, u, 16,  8,  uint, max)
+    OP_AARCH64_FUNC(max, s, 32,  4,   int, max)
+    OP_AARCH64_FUNC(max, u, 32,  4,  uint, max)
+#if defined(GENERATE_SVE_CODE)
+    OP_AARCH64_FUNC(max, s, 64,  2,   int, max)
+    OP_AARCH64_FUNC(max, u, 64,  2,  uint, max)
+#endif  /* defined(GENERATE_SVE_CODE) */
+
+    OP_AARCH64_FUNC(max, f, 32,  4, float, max)
```

Comment:
What? no love for half-precision floats?

Reply:
half-precision is yet in the MPI standard. I think there is (or was) a PR to add support for them in OMPI.